src/dataflow/src/render/upsert.rs

```diff
@@ -0,0 +1,271 @@
+
+    stream.unary_fallible(Pipeline, "UpsertLinearFallible", move |_, _| {
+        move |input, ok_output, err_output| {
+            input.for_each(|time, data| {
+                let mut ok_session = ok_output.session(&time);
+                let mut err_session = err_output.session(&time);
+                data.swap(&mut storage);
+                for (key, value, time) in storage.drain(..) {
+                    let temp_storage = RowArena::new();
+                    let datums = key.unpack();
+                    let key_pred_eval = key_predicates
+                        .iter()
+                        .map(|predicate| predicate.eval(&datums, &temp_storage))
+                        .find(|result| result != &Ok(Datum::True));
+                    match key_pred_eval {
+                        None => {
+                            if let Some(value) = value {
+                                let datums = value.unpack();
+                                let pred_eval = predicates
+                                    .iter()
+                                    .map(|predicate| predicate.eval(&datums, &temp_storage))
+                                    .find(|result| result != &Ok(Datum::True));
+                                match pred_eval {
+                                    None => {
+                                        let value = Some(row_packer.pack(position_or.iter().map(
+                                            |pos_or| match pos_or {
+                                                Some(index) => datums[*index],
+                                                None => Datum::Dummy,
+                                            },
+                                        )));
+                                        filtered_storage.push((key, value, time));
+                                    }
+                                    Some(Ok(Datum::False)) => {
+                                        filtered_storage.push((key, None, time))
+                                    }
+                                    Some(Ok(Datum::Null)) => {
+                                        filtered_storage.push((key, None, time))
+                                    }
+                                    Some(Ok(x)) => {
+                                        panic!("Predicate evaluated to invalid value: {:?}", x)
+                                    }
+                                    Some(Err(x)) => {
+                                        err_session.give(DataflowError::from(x));
+                                    }
+                                }
+                            } else {
+                                filtered_storage.push((key, None, time));
+                            }
+                        }
+                        Some(Ok(Datum::False)) | Some(Ok(Datum::Null)) => {}
+                        Some(Ok(x)) => panic!("Predicate evaluated to invalid value: {:?}", x),
```

Comment:
I'd love to turn these from panics into run-time errors. I understand that they are pervasive throughout the code, but it just occurred to me now and I wanted to flag it.

Reply:
So turn the panics into `DataflowError` to be handed off to the `err_session`?

Reply:
Can we sync up on this before we embark on this? I don't generally think that returning errors is the correct way to assert runtime invariants and surface programming errors, though given the design of timely/differential there is perhaps no other choice. But seems like it would be good to chat things out, maybe in a GH issue?

Reply:
Sorry, I should have been more clear that this isn't the PR to start that on. I think @awang has some plans to workshop ways to harden MZ so that it catches on fire in fewer places. That seems like a good place to talk through strategy.